src/commands/up.rs

```diff
@@ -646,6 +667,87 @@ fn trigger_command_for_resolved_app_source(
         .collect()
 }
 
+// Scrubs the locked app to only contain the given list of components
+// Introspects the LockedApp to find and selectively retain the triggers that correspond to those components
+fn retain_components(locked_app: &mut LockedApp, components: &[String]) -> Result<()> {
+    // Create a temporary app to access parsed component and trigger information
+    let tmp_app = spin_app::App::new("tmp", locked_app.clone());
+    let component_trigger_id_map = component_trigger_map(&tmp_app, components)?;
+    let components = component_trigger_id_map.keys().collect::<HashSet<_>>();
+    let triggers = component_trigger_id_map
+        .values()
+        .flatten()
+        .collect::<HashSet<_>>();
+    locked_app.components.retain(|c| components.contains(&c.id));
```

Comment:
Why collect `components` into a hashset instead of using [`component_trigger_id_map.contains_key`](https://doc.rust-lang.org/stable/std/collections/struct.HashMap.html#method.contains_key)?

Reply:
Good call out. This was legacy from a couple iterations ago.